plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConverter.kt

```diff
@@ -95,35 +61,38 @@ data class StmtConverter<out RTC : ResultTrackingContext>(
         methodCtx.addLoopIdentifier(targetName, whileIndex)
     }
 
-    override fun <R> withFreshWhile(action: StmtConversionContext<RTC>.() -> R): R {
-        val freshIndex = whileIndexProducer.getFresh()
-        val ctx = copy(whileIndex = freshIndex)
-        addDeclaration(ctx.continueLabel().toDecl())
-        addStatement(ctx.continueLabel().toStmt())
-        val result = ctx.action()
-        addDeclaration(ctx.breakLabel().toDecl())
-        addStatement(ctx.breakLabel().toStmt())
-        return result
-    }
+    override fun <R> withFreshWhile(label: FirLabel?, action: StmtConversionContext.() -> R): R =
+        withNewScopeImpl {
+            val freshIndex = whileIndexProducer.getFresh()
+            val ctx = copy(whileIndex = freshIndex)
+            label?.let { ctx.addLoopName(it.name) }
+            ctx.action()
+        }
 
-    override fun <R> withWhenSubject(subject: VariableEmbedding?, action: StmtConversionContext<RTC>.() -> R): R =
+    override fun <R> withWhenSubject(subject: VariableEmbedding?, action: StmtConversionContext.() -> R): R =
         copy(whenSubject = subject).action()
 
-    override fun <R> withCheckedSafeCallSubject(subject: ExpEmbedding?, action: StmtConversionContext<RTC>.() -> R): R =
+    override fun <R> withCheckedSafeCallSubject(subject: ExpEmbedding?, action: StmtConversionContext.() -> R): R =
         copy(checkedSafeCallSubject = subject).action()
 
-    override fun withCatches(
+    override fun <R> withCatches(
         catches: List<FirCatch>,
-        action: StmtConversionContext<RTC>.(catchBlockListData: CatchBlockListData) -> Unit,
-    ): CatchBlockListData {
+        action: StmtConversionContext.(catchBlockListData: CatchBlockListData) -> R,
+    ): Pair<CatchBlockListData, R> {
         val newCatchLabels = catches.map { Label(catchLabelNameProducer.getFresh(), listOf()) }
-        newCatchLabels.forEach { addDeclaration(it.toDecl()) }
         val exitLabel = Label(tryExitLabelNameProducer.getFresh(), listOf())
-        addDeclaration(exitLabel.toDecl())
         val ctx = copy(activeCatchLabels = activeCatchLabels + newCatchLabels)
         val catchBlockListData =
             CatchBlockListData(exitLabel, newCatchLabels.zip(catches).map { (label, firCatch) -> CatchBlockData(label, firCatch) })
-        ctx.action(catchBlockListData)
-        return catchBlockListData
+        val result = ctx.action(catchBlockListData)
+        return Pair(catchBlockListData, result)
+    }
+
+    private fun <R> withNewScopeImpl(action: StmtConverter.() -> R): R {
+        val newScopeIndex = scopeIndexProducer.getFresh()
+        val inner = copy(scopeIndex = newScopeIndex)
+        var result: R? = null
+        inner.withScopeImpl(newScopeIndex) { result = inner.action() }
+        return result!!
```

Comment:
This could be a good case where we can use Kotlin contracts for doing one-time initialization of the `result` value.

Reply:
I don't think that's possible since `withScopeImpl` is an interface function.

Reply:
Ah, that's unfortunate then.